### PR TITLE
[Hardware][Ascend] Integrate MindIE dense and sparse quantized attention for Wan2.2 T2V

### PR DESCRIPTION
**Pull request by De-cs**
## Purpose

Related RFC: https://github.com/vllm-project/vllm-omni/issues/7629

Enable quantized self-attention for Wan2.2 T2V A14B on Ascend through MindIE-SD:

| Attention path | Precision | Layer/step quantization skip |
| --- | --- | --- |
| Dense FA | MXFP8, MXFP4; existing FP8 dispatch also uses the public API | Floating-point Dense FA |
| BSA | FP8, MXFP4 | Floating-point BSA, retaining sparsity |

Omni selects the backend, precision and per-forward policy. MindIE-SD owns Q/K/V quantization, scale preparation, internal padding, mask selection/rearrangement for BSA and native operator execution. This PR contains no native C++ or custom-operator changes.

## Changes

- Delegate Dense quantized execution to `mindiesd.quant_attention`, preserving the caller's BSND/BNSD layout and softmax scale. Keep implicit BSND for existing callers and seeded FP8/MXFP8 rotations without changing global RNG. Remove the unused `fp8_rotate_quant_fa` compatibility wrapper and `is_quantized_kv_cache` helper.
- Pass BSA precision directly to `mindiesd.sparse_attention`: quantized single-video attention uses RFv3, and floating-point attention uses RFv2. Preserve MindIE's default BSA rotation. This PR does not depend on `get_bsa_supported_precisions` or any native precision-query operator.
- Extend the existing attention configuration with `quant.method`, ordered `quant.fallback`, FP8/MXFP8-only `quant.rotation_seed`, and per-role `quant.skip_layers`/`quant.skip_steps`. Preserve global skip selectors and legacy `block_sparse.precision`; reject conflicting settings and no-op rotation seeds for MXFP4/float.
- Apply layer/step skips before precision fallback. BSA precision fallback to `float` stays sparse. Existing sparse warmup, excluded layers and short-sequence rules can independently select Dense attention. Native execution errors propagate without retrying another precision or backend.
- Validate Dense and BSA input contracts separately before attempting execution. BSA requires identical full-sequence BSND Q/K/V; quantized Wan BSA is limited to batch one. Retain the Python API signature check because older `sparse_attention(**kwargs)` implementations silently ignore `precision`.
- Publish Wan's unpadded video grid and used length for sparse planning. Only masks explicitly identified by the model as alignment padding can be replaced by trimming; arbitrary masks and piecewise visibility stay on Dense attention. Retain Ulysses support and reject Ring/AllGather-KV configurations that do not provide the required full-sequence contract.
- Keep Wan cross-attention floating point. Recompute policy metadata each forward without mutating caller-owned metadata; skip unnecessary copies. Step indices span the complete request across both transformer stages, and layer indices are local to each transformer. Reject configured layer skips when the block index cannot be parsed.
- Provide four T2V deployment YAMLs under `examples/offline_inference/text_to_video/wan22_quant_attention/`. They use **40 steps**, `quant.skip_steps: "0,1,38,39"` and `quant.skip_layers: "0,39"`. These are fixed indices; adjust them when changing the step count.

The scope is Wan2.2 T2V attention integration. Weight/Linear quantization, I2V and new model support are outside this PR.

## Dependency requirements

- Official FP8/MXFP8 API baseline: MindIE-SD [`cf1a89be5803ad246f26c88c54db89a6a77748d4`](https://gitcode.com/Ascend/MindIE-SD/commit/cf1a89be5803ad246f26c88c54db89a6a77748d4).
- Required development snapshot for all four validation configurations: [`8637b5333b0225381b215390fd09a8732e671cc4`](https://gitcode.com/zqxu/MindIE-SD/commit/8637b5333b0225381b215390fd09a8732e671cc4). It provides the public APIs, explicit BSA `precision`/RFv3 support, and QFA native fixes. It is **not an upstream release**. Build the Python package, PyTorch plugin and custom operators from this exact checkout; copying Python files alone does not update native dependencies.
- Current upstream MXFP4 API proposal: [`9cbf9948325f15a6d9c5951389df564432b6d739`](https://gitcode.com/zqxu/MindIE-SD/commit/9cbf9948325f15a6d9c5951389df564432b6d739) in [MindIE-SD PR #630](https://gitcode.com/Ascend/MindIE-SD/pull/630). It does not contain the QFA native fixes and is not the four-path validation revision.

Dense MXFP4 remains experimental until an upstream dependency build passes device qualification. The original QFA build had failures at Wan valid length 75600 / padded length 75776. An exported Python symbol, finite outputs alone, or a saved video followed by process abort do not qualify the dependency. Record source/wheel fingerprints, loaded libraries, device, CANN and torch/torch_npu versions. Use the vLLM version required by this Omni checkout; earlier vLLM 0.28 development runs do not qualify this branch.

## Test plan

From the repository root in a configured Omni environment with compatible MindIE-SD:

```bash
python -m pytest \
    tests/diffusion/attention/test_mindie_runtime.py \
    tests/diffusion/attention/test_mindie_public_integration.py \
    tests/diffusion/attention/test_rainfusion_plan.py \
    tests/diffusion/models/wan2_2/test_wan22_mindie_metadata.py \
    tests/platforms/npu/quant/test_kv_quant_npu.py \
    -k 'not real_npu' -v -o addopts=''
```

The cross-repository tests execute real MindIE Python public adapters with native operators substituted. They cover quantization axes, layout, scales, sequence metadata, padding and output restoration; these are interface tests, not numerical quantization or quality tests.

For native Dense qualification on a supported NPU with the matching dependency build:

```bash
python -m pytest tests/platforms/npu/quant/test_kv_quant_npu.py \
    -k real_npu -vv -s -o addopts=''
```

Native cases cover FP8/MXFP8/MXFP4, BSND/BNSD, D=128 and lengths 256/75600, including per-head constant-V numerical references and tail restoration. Skipped cases are not passes.

The [quantized attention guide](https://github.com/De-cs/vllm-omni/blob/codex/wan22-dense-attention/docs/user_guide/quantization/quantized_kvcache.md#npu-validation) provides commands for all four deployment configurations. Repeat with 81 frames at 720p for full geometry. For BSA, confirm that the real generation selects sparse attention rather than the short-sequence Dense path. Require the intended precision, a decodable video and process exit 0. VBench Standard quality and controlled performance comparisons are separate validation.

## Validation status

Reviewed Omni revision: `3ab6a9ca590862cc53f45faf64b86d2c6a46e872`.

This combined update passed **121 CPU contract tests**, with **14 hardware-dependent cases skipped**. The isolated Linux CPU bootstrap substitutes unavailable vLLM/platform imports and native operators; real Omni method bodies and MindIE Python public adapters execute. It is not a full installed-package, NPU or model end-to-end run. Coverage includes all four target paths, legacy global FP8, configured precision fallback, layer/step skips, both transformer stages, consecutive-request isolation, caller masks, metadata ownership and execution errors without retries.

Ruff 0.14.10 passed for all changed Python files; four deployment YAML checks, Markdown checks, and documented Bash syntax checks passed. Mypy 3.10 passed for the changed test, configuration, and sparse backend files. Library checking still reports three existing `attention/layer.py` errors, also reproduced using the same seven files on unmodified base `85da94339130337b49e610d08276aedcf84f6b96`; this PR does not claim a clean whole-library type check.

NPU execution, model generation, VBench Standard quality and performance measurements have **not been rerun for this combined review update**. Earlier development smoke runs used separate dependency builds and are not presented as validation of this PR revision. No quality or speedup claim is made.
### Visual comparison

All videos use the same generation settings:

- Model: Wan2.2 T2V A14B
- Prompt: `<PROMPT>`
- Seed: `42`
- Resolution: `1280x720`
- Frames: `81`
- Denoising steps: `50`
- Weight precision: BF16
- Attention fallback: `skip_layers="0,39"`, `skip_steps="0,1,48,49"`

> These videos provide functional and visual comparison only. VBench results
> will be added separately.

| Case | Backend | Attention precision | Result |
| --- | --- | --- | --- |
| BF16 baseline | `FLASH_ATTN` | BF16 | ✅ Generated successfully |
| Dense MXFP8 | `FLASH_ATTN` | MXFP8 | ✅ Generated successfully |
| Dense MXFP4 | `FLASH_ATTN` | MXFP4 | ✅ Generated successfully |
| BSA FP8 | `RAINFUSION_ATTN` | FP8 | ✅ Generated successfully |
| BSA MXFP4 | `RAINFUSION_ATTN` | MXFP4 | ✅ Generated successfully |

#### BF16 baseline

<details>
<summary><b>BF16 </b></summary>

[BF16_VIDEO_URL](https://github.com/user-attachments/assets/d038f390-140b-40b2-81dc-aa89a3090312)
</details>

#### Dense Flash Attention

<details>
<summary><b>Dense MXFP8</b></summary>

[DENSE_MXFP8_VIDEO_URL](https://github.com/user-attachments/assets/2d8ada26-8657-4fd9-a828-1f18773b4ba9)

</details>

<details>
<summary><b>Dense MXFP4</b></summary>

[DENSE_MXFP4_VIDEO_URL](https://github.com/user-attachments/assets/15fc8f41-a49d-4b3d-b293-6328a56c66f4)

</details>

#### Block-sparse Attention

<details>
<summary><b>BSA FP8</b></summary>

[BSA_FP8_VIDEO_URL](https://github.com/user-attachments/assets/5dcf3dab-c837-422d-956c-f55035585eee)

</details>

<details>
<summary><b>BSA MXFP4</b></summary>

[BSA_MXFP4_VIDEO_URL](https://github.com/user-attachments/assets/b37f554f-87e9-43d6-a7f1-9ae22deb1786)

</details>
This change was developed with AI assistance.

### CODE REVIEW

File: vllm_omni/diffusion/attention/backends/flash_attn.py [hunk 1]
```diff
@@ -526,30 +531,103 @@ def forward_npu(
...
+    def _quant_unsupported_reason(
+        self,
+        method: str,
+        query: torch.Tensor,
+        key: torch.Tensor,
+        value: torch.Tensor,
+        attn_metadata: AttentionMetadata | None,
+    ) -> str | None:
+        layout = self.qkv_layout
+        if layout not in ("BSND", "BNSD"):
```
Comment (lishunyang12) [inline vllm_omni/diffusion/attention/backends/flash_attn.py]: Preserve implicit BSND for existing callers, including QwenImage's global FP8 path.

File: vllm_omni/diffusion/attention/layer.py [hunk 2]
```diff
@@ -335,18 +346,21 @@ def _should_apply_kv_cache_quant(self) -> bool:
...
-        if kv_cache_dtype is None or self._disable_kv_quant or not self._should_apply_kv_cache_quant():
-            if attn_metadata is None or "kv_cache_dtype" not in attn_metadata.extra:
-                return attn_metadata
-            extra = dict(attn_metadata.extra)
-            extra.pop("kv_cache_dtype", None)
-            return replace(attn_metadata, extra=extra)
-
+        disabled = self._disable_kv_quant or not self._should_apply_kv_cache_quant()
+        dtype = self._kv_cache_dtype
+        extra = dict(attn_metadata.extra) if attn_metadata is not None else {}
```
Comment (lishunyang12) [inline vllm_omni/diffusion/attention/layer.py:365]: Skip metadata copies when neither quantization policy nor stale quantization keys require changes.

File: docs/user_guide/quantization/quantized_kvcache.md [hunk 3]
```diff
@@ -123,3 +123,21 @@ layers skip FP8 FA; all other eligible full-attention forwards use the FP8 path.
 4. Report both latency and quality results when enabling this option for a new
    model. For image or video models, include visual comparison and quantitative
    metrics when available, such as PSNR or SSIM.
+
+## Wan2.2 T2V quantized attention on Ascend
+
+For Wan2.2 T2V A14B, `FLASH_ATTN` accepts per-role `quant.method` values
+`fp8`, `mxfp8`, `mxfp4`, or `float`. This requires a compatible MindIE-SD
```
Comment (lishunyang12) [inline docs/user_guide/quantization/quantized_kvcache.md]: Pin the required MindIE-SD revision and add runnable NPU validation commands.

File: vllm_omni/diffusion/data.py [hunk 4]
```diff
@@ -1802,10 +1802,34 @@ class AttnQuantSpec:
...
+            if self.q_block_size != 1 or self.k_block_size != 16:
+                raise ValueError("quant.method uses Runtime block sizes; do not set q_block_size/k_block_size.")
+        elif self.fallback or self.rotation_seed is not None:
+            raise ValueError("fallback and rotation_seed require quant.method.")
+        if not isinstance(self.fallback, list) or any(m not in self._VALID_METHODS for m in self.fallback):
+            raise ValueError("quant.fallback must be a list of float/fp8/mxfp8/mxfp4 methods.")
+        chain = ([self.method] if self.method is not None else []) + self.fallback
+        if len(chain) != len(set(chain)) or ("float" in chain and chain[-1] != "float"):
+            raise ValueError("Quantization methods must be unique and float must be the last fallback.")
+        if self.rotation_seed is not None and (
```
Comment (david6666666) [inline vllm_omni/diffusion/data.py]: [P2] Reject rotation_seed for non-rotating methods — mxfp4 and float currently accept it, but dispatch only forwards the seed for fp8/mxfp8, so these configurations silently do nothing. Please validate it as FP8/MXFP8-only or make the no-op explicit.

**Comment by david6666666** [comment]
please give video output comparasion, and @lyj-jjj @fan2956 ptal thx

File: vllm_omni/diffusion/attention/layer.py [hunk 5]
```diff
@@ -321,32 +321,33 @@ def _init_kv_cache_quantization(self, config) -> None:
...
-
+        disabled = self._disable_kv_quant or not self._should_apply_kv_cache_quant()
+        dtype = "float" if disabled or self._kv_cache_dtype == "float" else self._kv_cache_dtype
+        if dtype is None and (attn_metadata is None or "kv_cache_dtype" not in attn_metadata.extra):
+            return attn_metadata
+        extra = dict(attn_metadata.extra) if attn_metadata is not None else {}
+        # Recompute per forward so shared metadata cannot retain another step's policy.
+        extra.pop("kv_cache_dtype", None)
+        if dtype is not None:
+            extra["kv_cache_dtype"] = dtype
```
Comment (david6666666) [inline vllm_omni/diffusion/attention/layer.py:369]: [P2] Keep the float marker out of non-NPU metadata — When no diffusion quantization is configured, `disable_kv_quant=True` still makes this write `kv_cache_dtype="float"`. CUDA then sees a non-empty `extra` dict and skips the existing automatic FP32→SDPA fallback in `_run_local_attention`. Please only publish this marker when an NPU quantization policy is active, or exclude it from the non-NPU fallback predicate.

Comment (De-cs) [reply vllm_omni/diffusion/attention/layer.py:369]: Fixed in `5433cf59`.

`_with_kv_cache_dtype` now publishes `kv_cache_dtype="float"` only when a non-float quantization policy is active and the current layer or denoising step opts out of quantization. When the configured dtype is `None`, `auto`, or `float`, no quantization marker is added, so CUDA FP32 attention remains eligible for the existing automatic SDPA fallback.

I also added a regression test covering these three no-quantization configurations while preserving the float fallback marker for layer/step skips under an active quantization policy.

File: vllm_omni/diffusion/data.py (hunk 4 above)
Comment (De-cs) [reply vllm_omni/diffusion/data.py]: Addressed. I removed `rotation_seed` from the public attention quantization configuration entirely. Dense FP8/MXFP8 now use Omni's existing deterministic cached rotation, while MXFP4 and floating-point execution no longer accept an option that would be silently ignored.

File: docs/user_guide/quantization/quantized_kvcache.md (hunk 3 above)
Comment (De-cs) [reply docs/user_guide/quantization/quantized_kvcache.md]: Addressed. The documentation now pins MindIE-SD revision `8637b5333b0225381b215390fd09a8732e671cc4` and clarifies that the Python package, PyTorch plugin, and custom operators must be built from the same checkout. It also provides a copy-pasteable real-NPU pytest command. The test matrix covers FP8, MXFP8, and MXFP4 with both BSND and BNSD layouts, D=128, and the Wan sequence length S=75600; skipped NPU tests are explicitly not treated as passing.

File: vllm_omni/diffusion/attention/layer.py (hunk 2 above)
Comment (De-cs) [reply vllm_omni/diffusion/attention/layer.py:365]: Addressed. `_with_kv_cache_dtype` now returns the original metadata unchanged when no quantization dtype is configured and there is no stale `kv_cache_dtype` key. A metadata copy is created only when the per-forward policy needs to add, replace, or remove that key.

File: vllm_omni/diffusion/attention/backends/flash_attn.py (hunk 1 above)
Comment (De-cs) [reply vllm_omni/diffusion/attention/backends/flash_attn.py]: Addressed. The quantized NPU path now resolves the layout with `self.qkv_layout or "BSND"`, preserving the implicit BSND behavior for existing callers. The regression test `test_legacy_global_fp8_uses_implicit_bsnd` also verifies that the legacy global FP8 path passes `layout="BSND"` to MindIE-SD when no layout is explicitly configured.

**Comment by jiangmengyu18** [comment]
LGTM.

**Review by david6666666 — APPROVED** [review]

**Comment by fan2956** [comment]
LGTM
